### DISCUSSION TITLE
Self-collisions & mutiple-object collisions

### DISCUSSION THREAD

**Discussion by Ch0veS0uris** [evalf/nutils#886] [Q&A]
Hello,

1. Is there some ways (or it is planned) to add self-collisions detection for a deformable object? Like for the torsion.py example, but with a rotation of 360°?

1. It is possible (or planned) to simulate multiple objects (rigid or non-rigid) with collisions?

Best

**Comment by gertjanvanzwieten** [comment]
Physical collisions are not planned as a built-in feature, as Nutils operates at the slightly lower level of numerical toolbox. But it does have some instruments that may be useful in the implementation of a collision strategy, such as [locate](https://docs.nutils.org/en/v8.0/nutils/topology/#nutils.topology.Topology.locate). Multiple objects would involve maintaining multiple independent topologies, but you would need to program a strategy for these topologies to interact.

**Reply by Ch0veS0uris** [reply]
Thank you very much @gertjanvanzwieten. In addition, for one object (2D or 3D), could we define several domains (or "patches") with different materials properties (young modulus, Poisson coefficient, or thermal parameters)?

**Reply by gertjanvanzwieten** [reply]
Certainly! You can do this either by splitting your integral into several integrals to differentiate the integrands, or use your topology's `indicator` method to form parameters that take different values over the subdomains, so you can still write your weak form as a single integral. I notice that `indicator` lacks documentation, but it takes a single argument, either the name of the subdomain or the subdomain as a topology, and returns a function that evaluates to 1 on the subdomain and 0 outside of it.

Edit: in case your question pertains to the subdomains themselves, these can be defined using slice operations if your topology is structured (see the coil example), or using the `select` method (see adaptivity, note that the selection is element-wise) or if you use gmsh via physical groups in the .geo file.

**Reply by Ch0veS0uris** [reply]
Thanks again @gertjanvanzwieten for the tips :) . My question is indeed to define the subdomains on a 3D deformable object. So it would also be great to have a variation of the examples (platewithhole-nurbs.py or/and torsion.py) without hole, but rather with the materials properties defined for each subdomain.

**Reply by gertjanvanzwieten** [reply]
We're working towards a new release so that's absorbing time but I'll see if we can add an example of that kind, I agree it would be useful.
